keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java

```diff
@@ -0,0 +1,202 @@
+
+        Assert.assertArrayEquals(new long[]{3, 2, 1, 2, 3}, forwardWrtA.getShape());
+        Assert.assertArrayEquals(new long[]{3, 2, 1, 2, 3}, backwardWrtA.getShape());
+        Assert.assertArrayEquals(forwardWrtA.asFlatDoubleArray(), backwardWrtA.asFlatDoubleArray(), 1e-6);
+
+    }
+
+    @Test
+    public void canPartialCorrectlyFlowThroughSumThenPermute() {
+        UniformVertex A = new UniformVertex(0, 10);
+        A.setValue(DoubleTensor.arange(0, 6).reshape(1, 2, 3));
+
+        UniformVertex B = new UniformVertex(0, 10);
+        B.setValue(DoubleTensor.arange(0, 6).reshape(1, 2, 3));
+
+        DoubleVertex C = A.plus(B);
+
+        DoubleVertex sum = C.sum(2);
+
+        PermuteVertex permute = sum.permute(1, 0);
+
+        DoubleTensor forwardWrtA = Differentiator.forwardModeAutoDiff(A, permute).of(permute);
+        DoubleTensor backwardWrtA = Differentiator.reverseModeAutoDiff(permute, ImmutableSet.of(A)).withRespectTo(A);
+
+        Assert.assertArrayEquals(new long[]{2, 1, 1, 2, 3}, forwardWrtA.getShape());
+        Assert.assertArrayEquals(new long[]{2, 1, 1, 2, 3}, backwardWrtA.getShape());
+        Assert.assertArrayEquals(forwardWrtA.asFlatDoubleArray(), backwardWrtA.asFlatDoubleArray(), 1e-6);
+
+    }
+
+    @Test
+    public void canCalculateAutoDiffOfGraphWithRankTwoPermute() {
+        UniformVertex A = new UniformVertex(0, 10);
+        A.setValue(DoubleTensor.arange(0, 2).reshape(2, 1));
+
+        UniformVertex B = new UniformVertex(0, 10);
+        B.setValue(DoubleTensor.arange(0, 2).reshape(2, 1));
+
+        MultiplicationVertex C = A.times(B);
+
+        PermuteVertex permute = C.permute(1, 0);
+
+        UniformVertex E = new UniformVertex(0, 10);
+        E.setValue(DoubleTensor.arange(0, 2).reshape(1, 2));
+
+        MultiplicationVertex F = permute.times(E);
+
+        DoubleTensor forwardWrtA = Differentiator.forwardModeAutoDiff(A, F).of(F);
+        DoubleTensor backwardWrtA = Differentiator.reverseModeAutoDiff(F, ImmutableSet.of(A)).withRespectTo(A);
+
+        Assert.assertArrayEquals(new long[]{1, 2, 2, 1}, forwardWrtA.getShape());
+        Assert.assertArrayEquals(new long[]{1, 2, 2, 1}, backwardWrtA.getShape());
+        Assert.assertArrayEquals(forwardWrtA.asFlatDoubleArray(), backwardWrtA.asFlatDoubleArray(), 1e-6);
+    }
+
+    @Test
+    public void autoDiffIsUnchangedWhenPermuteIsReversed() {
+        UniformVertex A = new UniformVertex(0, 10);
+        A.setValue(DoubleTensor.arange(0, 6).reshape(1, 2, 3));
+
+        UniformVertex B = new UniformVertex(0, 10);
+        B.setValue(DoubleTensor.arange(0, 6).reshape(1, 2, 3));
+
+        MultiplicationVertex C = A.times(B);
+
+        PermuteVertex permute = C.permute(2, 0, 1);
+        PermuteVertex revertThePermute = permute.permute(1, 2, 0);
+
+        Assert.assertArrayEquals(C.getValue().asFlatDoubleArray(), revertThePermute.getValue().asFlatDoubleArray(), 1e-6);
```

Comment:
Comparing only the flat arrays doesn't check for shape issues. You can use one of Mike's hamcrest matchers or assert that the two tensors are equal.

Reply:
Changed to be a matcher